python/paddle/tensor/stat.py

```diff
@@ -253,6 +253,140 @@ def numel(x, name=None):
+            For more information, please refer to :ref:`api_guide_Name`.
+
+    Returns:
+        Tensor, results of median along ``axis`` of ``x``. The output dtype is the same as `x`.
+
+    Examples:
+        .. code-block:: python
+            :name: nanmedian-example
+
+            import paddle
+            import numpy as np
+            x = paddle.to_tensor([[np.nan, 2. , 3. ], [0. , 1. , 2. ]])
+
+            y1 = x.nanmedian()
+            # y1 is [[[2.]]
+
+            y2 = x.nanmedian(0)
+            # y2 is [[0.,  1.5, 2.5]]
+
+            y3 = x.nanmedian(0, keepdim=False)
+            # y3 is [0.,  1.5, 2.5]
+    """
+    if not isinstance(x, Variable):
+        raise TypeError("In median, the input x should be a Tensor.")
+
+    dims = len(x.shape)
+    out_shape = list(x.shape)
+    if axis is None:
+        x = paddle.flatten(x)
+        axis = 0
+        out_shape = [1] * dims if keepdim else [1]
+    else:
+        if isinstance(axis, tuple):
+            axis = list(axis)
+
+        if isinstance(axis, list):
+            for k in range(len(axis)):
+                if axis[k] < 0:
+                    axis[k] += dims
+            axis = list(set(axis))
+            if len(axis) <= 0:
+                raise ValueError("axis should not be empty")
+
+            axis_src, axis_dst = [], []
+            for axis_single in axis:
+                if not isinstance(axis_single, int) or not (
+                        axis_single < dims and axis_single >= -dims):
+                    raise ValueError(
+                        "Axis should be None, int, or a list, element should in range [-rank(x), rank(x))."
+                    )
+                axis_src.append(axis_single)
+                out_shape[axis_single] = 1 if keepdim else 0
+
+            axis_dst = list(range(-len(axis), 0))
+            x = paddle.moveaxis(x, axis_src, axis_dst)
+            x = paddle.flatten(x, axis_dst[0], axis_dst[-1])
+            axis = axis_dst[0]
+        else:
+            if not isinstance(axis, int) or not (axis < dims and axis >= -dims):
+                raise ValueError(
+                    "Axis should be None, int, or a list, element should in range [-rank(x), rank(x))."
+                )
+            if axis < 0:
+                axis += dims
+            out_shape[axis] = 1 if keepdim else 0
+
+    trans_axis = []
+    last_axis = len(x.shape) - 1
+    if len(x.shape) > 1 and axis is not None and axis != last_axis:
+        for i in range(axis):
+            trans_axis.append(i)
+        trans_axis.append(last_axis)
+        for i in range(axis + 1, last_axis):
+            trans_axis.append(i)
+        trans_axis.append(axis)
+        x = paddle.transpose(x, perm=trans_axis)
+
+    if len(x.shape) == 1 and axis == 0:
+        axis = None
+    else:
+        axis = len(x.shape) - 1
+    res_shape = [x for x in out_shape if x > 0]
+
+    if _in_legacy_dygraph():
+        medians, out = _C_ops.nanmedian(x, 'ignore_nan', ignore_nan)
```

Comment:
According to the comments above, the parameters of Python API and C++ API should be the same, and the preprocessing logic in Python above should also be put into C++.

Reply:
A question: can c++ api support the python data type like tuple and None?

Reply:
done